### PR TITLE
Socket

### PR DESCRIPTION
**Pull request by arhtudormorar**

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -76,7 +76,7 @@
     ],
     "@typescript-eslint/indent": "off",
     "@typescript-eslint/explicit-module-boundary-types": "off",
-    "@typescript-eslint/no-use-before-define": ["error", {"functions": false, "classes": false}],
+    "@typescript-eslint/no-use-before-define": ["off", {"functions": false, "classes": false}],
```
Comment (StanislavSava) [inline .eslintrc:79]: what is the reason for this change?

File: package.json [hunk 2]
```diff
@@ -113,7 +113,7 @@
     "@elrondnetwork/erdjs": "9.1.0",
     "@reduxjs/toolkit": "1.6.2",
     "@testing-library/jest-dom": "5.16.2",
-    "@typescript-eslint/parser": "5.22.0",
+    "@typescript-eslint/parser": "4.4.1",
```
Comment (StanislavSava) [inline package.json:116]: Why the downgrade?

File: src/apiCalls/transactions/getTransactionsByHashes.ts [hunk 3]
```diff
@@ -50,7 +50,7 @@ export async function getTransactionsByHashes(
       receiver: txOnNetwork?.receiver,
       pendingResults: txOnNetwork.pendingResults,
       previousStatus,
-      hasStatusChanged: status !== previousStatus
+      hasStatusChanged: txOnNetwork.status !== previousStatus
```
Comment (StanislavSava) [inline src/apiCalls/transactions/getTransactionsByHashes.ts:53]: Why the change?

File: src/components/InternalComponents.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import React, { Component } from 'react';
+import TransactionSender from './TransactionSender';
+import TransactionsTracker from './TransactionsTracker';
+
+export class InternalComponents extends Component {
```
Comment (StanislavSava) [inline src/components/InternalComponents.tsx:5]: Do we still need these?

File: src/components/TransactionSender/index.ts [hunk 5]
```diff
@@ -18,12 +19,17 @@ import {
 } from 'types/enums';
 import { setNonce } from 'utils';
 
-const TransactionSender = () => {
+export interface TransactionSenderType {
+  sendSignedTransactionsAsync?: SendSignedTransactionsAsyncType;
+}
+
+export const TransactionSender = ({
+  sendSignedTransactionsAsync: customSendTxAsync
```
Comment (StanislavSava) [inline src/components/TransactionSender/index.ts:27]: why the name change? can't we use the default name, sendSignedTransactionsAsync?

File: src/components/TransactionSender/index.ts [hunk 6]
```diff
@@ -18,12 +19,17 @@ import {
...
+}: TransactionSenderType) => {
   const account = useSelector(accountSelector);
   const signedTransactions = useSelector(signedTransactionsSelector);
-  const { sendSignedTransactionsAsync } = useContext(
-    OverrideDefaultBehaviourContext
-  );
+
+  const sendSignedTransactionsAsync = customSendTxAsync ?? defaultSendSignedTxs;
```
Comment (StanislavSava) [inline src/components/TransactionSender/index.ts:32]: ?? will fail, due to microbundle, use || instead

File: src/components/TransactionSender/index.ts [hunk 7]
```diff
@@ -1,8 +1,9 @@
-import React, { useContext, useRef } from 'react';
+import React, { useRef } from 'react';
 import { Address } from '@elrondnetwork/erdjs';
 import { Signature } from '@elrondnetwork/erdjs/out/signature';
 
-import OverrideDefaultBehaviourContext from 'contexts/OverrideDefaultBehaviourContext';
+import { sendSignedTransactions as defaultSendSignedTxs } from 'apiCalls/transactions';
```
Comment (StanislavSava) [inline src/components/TransactionSender/index.ts:5]: Since this is reused, it should probably be moved to types folder

File: src/components/TransactionsTracker/useTransactionsTracker.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import { useEffect, useRef } from 'react';
+import { getTransactionsByHashes as defaultGetTxByHash } from 'apiCalls/transactions';
+import { GetTransactionsByHashesType } from 'contexts/types';
+import { useCheckTransactionStatus } from 'hooks';
+import { useGetPendingTransactions } from 'services/transactions';
+import { getIsTransactionPending } from 'utils';
+
+export interface TransactionsTrackerType {
```
Comment (StanislavSava) [inline src/components/TransactionsTracker/useTransactionsTracker.ts:8]: Check the file above, same interface, declared twice. Probably should be lifted up or moved to types folder

File: src/components/TransactionsTracker/useTransactionsTracker.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  getTransactionsByHash?: GetTransactionsByHashesType;
+}
+
+export function useTransactionsTracker(props?: TransactionsTrackerType) {
+  const { pendingTransactionsArray } = useGetPendingTransactions();
+
+  const getTransactionsByHash =
+    props?.getTransactionsByHash ?? defaultGetTxByHash;
```
Comment (StanislavSava) [inline src/components/TransactionsTracker/useTransactionsTracker.ts:16]: There is already a fallback on the default, either leave it only here it only in the component above

File: src/components/TransactionsTracker/useTransactionsTracker.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    ([sessionId, transactionBatch]) => {
+      const isPending =
+        sessionId != null && getIsTransactionPending(transactionBatch.status);
+      return isPending;
+    }
+  );
+
+  useEffect(() => {
+    if (pendingBatches.length > 0) {
+      intervalRef.current = setInterval(() => {
```
Comment (StanislavSava) [inline src/components/TransactionsTracker/useTransactionsTracker.ts:31]: This should be used only if the behaviour is not overwritten, otherwise we'll end up with an interval, regardless of the way we check for tx status

File: src/components/index.ts [hunk 11]
```diff
@@ -0,0 +1 @@
+export * from './InternalComponents';
```
Comment (StanislavSava) [inline src/components/index.ts:1]: Do we still need this?

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+} from 'types/transactions';
+import { getIsTransactionFailed, getIsTransactionPending } from 'utils';
+import { refreshAccount } from 'utils/account';
+import getPendingTransactions from './getPendingTransactions';
+
+interface TransactionStatusTrackerPropsType {
+  sessionId: string;
+  transactionBatch: SignedTransactionsBodyType;
+  getTransactionsByHash?: GetTransactionsByHashesType;
+  refetchTimestamp?: number;
```
Comment (StanislavSava) [inline src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts:26]: obsolete

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+interface RetriesType {
+  [hash: string]: number;
+}
+
+function useManageFailed() {
+  const dispatch = useDispatch();
```
Comment (StanislavSava) [inline src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts:34]: This seems to be a hook only because of useDispatch, it will probably be cleaner to just use store.dispatch and make this a regular function

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+      for (const serverTransaction of serverTransactions) {
+        const {
+          hash,
+          status,
+          results,
+          invalidTransaction,
+          pendingResults,
+          hasStatusChanged
+        } = serverTransaction;
+        try {
```
Comment (StanislavSava) [inline src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts:138]: I know this is my code, by would you mind moving this to a separate function? Something like "checkPendingTransactionStatus(serverTransaction)"

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+          }
+
+          if (invalidTransaction || getIsTransactionPending(status)) {
+            retriesRef.current[hash] = retriesRef.current[hash]
+              ? retriesRef.current[hash] + 1
+              : 1;
+            return;
+          }
+
+          if (!pendingResults) {
```
Comment (StanislavSava) [inline src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts:153]: pending results doesn't exist anymore, a transaction will enter success state only after the sc call executes, we can drop this

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+            timeoutRefs.current.push(hash);
+
+            const transitionToCompletedDelay =
+              customTransactionInformation?.completedTransactionsDelay || 0;
+            setTimeout(
```
Comment (StanislavSava) [inline src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts:158]: We should do a little change here. Check if the status is success "getIsTrnasactionsSuccess), and only then apply the delay. if the transaction has failed, the dispatch should happen immediately

**Review by StanislavSava — CHANGES_REQUESTED** [review]

File: src/apiCalls/transactions/getTransactionsByHashes.ts (hunk 3 above)
Comment (arhtudormorar) [reply src/apiCalls/transactions/getTransactionsByHashes.ts:53]: status was not defined. TBD

File: src/components/TransactionSender/index.ts (hunk 5 above)
Comment (arhtudormorar) [reply src/components/TransactionSender/index.ts:27]: sendSignedTransactionsAsync is used below, name change is to differentiate between default and custom

File: src/components/TransactionSender/index.ts (hunk 6 above)
Comment (arhtudormorar) [reply src/components/TransactionSender/index.ts:32]: dapp-core is now on `tsdx`, and working ✅

File: src/components/TransactionSender/index.ts (hunk 7 above)
Comment (arhtudormorar) [reply src/components/TransactionSender/index.ts:5]: it's a function. TBD

File: src/components/TransactionsTracker/index.tsx [hunk 17]
```diff
@@ -1,22 +1,19 @@
...
-    ([sessionId, transactionPayload]) => (
-      <TransactionStatusTracker
-        key={sessionId}
-        sessionId={sessionId}
-        transactionPayload={transactionPayload as SignedTransactionsBodyType}
-      />
-    )
-  );
+export interface TransactionsTrackerType {
+  refetchTimestamp?: number;
```
[the remark this answers is not in the document]
Comment (arhtudormorar) [reply src/components/TransactionsTracker/index.tsx:7]: ✅

File: src/components/TransactionsTracker/useTransactionsTracker.ts (hunk 8 above)
Comment (arhtudormorar) [reply src/components/TransactionsTracker/useTransactionsTracker.ts:8]: ✅

File: src/components/TransactionsTracker/useTransactionsTracker.ts (hunk 9 above)
Comment (arhtudormorar) [reply src/components/TransactionsTracker/useTransactionsTracker.ts:16]: ✅

File: src/components/index.ts (hunk 11 above)
Comment (arhtudormorar) [reply src/components/index.ts:1]: yes, for default. TBD

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts (hunk 13 above)
Comment (arhtudormorar) [reply src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts:34]: ✅

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+      });
+    }
+  }
+
+  async function checkBatch({
+    sessionId,
+    transactionBatch: { transactions, status, customTransactionInformation },
+    getTransactionsByHash = defaultGetTxByHash
+  }: TransactionStatusTrackerPropsType) {
+    const isPending = sessionId != null && getIsTransactionPending(status);
```
[the remark this answers is not in the document]
Comment (arhtudormorar) [reply src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts:112]: ✅

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+        status: TransactionBatchStatusesEnum.fail,
+        errorMessage: resultWithError?.returnMessage
+      })
+    );
+  };
+}
+
+function useManageFailedTransactions() {
+  const dispatch = useDispatch();
+  return (sessionId: string) => {
```
[the remark this answers is not in the document]
Comment (arhtudormorar) [reply src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts:68]: ✅

File: src/index.tsx [hunk 20]
```diff
@@ -12,6 +12,7 @@ export { AuthenticatedRoutesWrapper, AppInitializer } from './wrappers';
 export * as transactionServices from './services/transactions';
 export * as constants from './constants/index';
 export * as models from './models';
+export * as DappComponents from './components';
```
[the remark this answers is not in the document]
Comment (arhtudormorar) [reply src/index.tsx:15]: yes, for usage in Wallet. TBD

File: package.json (hunk 2 above)
Comment (arhtudormorar) [reply package.json:116]: ✅

File: src/components/InternalComponents.tsx (hunk 4 above)
Comment (arhtudormorar) [reply src/components/InternalComponents.tsx:5]: yes. TBD

File: src/components/TransactionSender/index.ts (hunk 5 above)
Comment (arhtudormorar) [reply src/components/TransactionSender/index.ts:27]: TBD

File: src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts (hunk 12 above)
Comment (arhtudormorar) [reply src/hooks/transactions/useCheckTransactionStatus/useCheckTransactionStatus.ts]: ✅

File: .eslintrc (hunk 1 above)
Comment (arhtudormorar) [reply .eslintrc:79]: ✅

File: src/components/TransactionsTracker/useTransactionsTracker.ts [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+
+  const getTransactionsByHash =
+    props?.getTransactionsByHash ?? defaultGetTxByHash;
+
+  const checkTransactionStatus = useCheckTransactionStatus();
+  const intervalRef = useRef<ReturnType<typeof setInterval>>();
+
+  const pendingBatches = pendingTransactionsArray.filter(
+    ([sessionId, transactionBatch]) => {
+      const isPending =
```
Comment (StanislavSava) [inline src/components/TransactionsTracker/useTransactionsTracker.ts:23]: I would rename this to "isBatchPending" :D
